File: modules/services_discovery/freebox.c
/*****************************************************************************
 * freebox.c:  Shoutcast services discovery module
 *****************************************************************************
 * Copyright (C) 2005-2007 the VideoLAN team
 * $Id$
 *
 * Authors: Sigmund Augdal Helberg <dnumgis@videolan.org>
 *          Antoine Cellerier <dionoea -@T- videolan -d.t- org>
 *          Pierre d'Herbemont <pdherbemont # videolan.org>
 *
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 2 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software
 * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston MA 02110-1301, USA.
 *****************************************************************************/

/*****************************************************************************
 * Includes
 *****************************************************************************/

#ifdef HAVE_CONFIG_H
# include "config.h"
#endif

#include <vlc_common.h>
#include <vlc_plugin.h>
#include <vlc_services_discovery.h>

/*****************************************************************************
 * Module descriptor
 *****************************************************************************/

static int  Open( vlc_object_t * );
static void Close( vlc_object_t * );

static int vlc_sd_probe_Open( vlc_object_t * );

vlc_module_begin ()
    set_category( CAT_PLAYLIST )
    set_subcategory( SUBCAT_PLAYLIST_SD )

    set_shortname( "Freebox")
    set_description( N_("Freebox TV") )
    set_capability( "services_discovery", 0 )
    set_callbacks( Open, Close )
    add_shortcut( "freebox" )

    VLC_SD_PROBE_SUBMODULE
vlc_module_end ()


/*****************************************************************************
 * Local prototypes
 *****************************************************************************/

static void *Run( void * );
struct services_discovery_sys_t
{
    vlc_thread_t thread;
};

/*****************************************************************************
 * Open: initialize and create stuff
 *****************************************************************************/
static int Open(vlc_object_t *this)
{
    services_discovery_t *sd = (services_discovery_t *)this;
    sd->p_sys = malloc(sizeof(*(sd->p_sys)));
    if (sd->p_sys == NULL)
        return VLC_ENOMEM;

    if (vlc_clone(&sd->p_sys->thread, Run, sd, VLC_THREAD_PRIORITY_LOW))
    {
        free(sd->p_sys);
        return VLC_EGENERIC;
    }
    return VLC_SUCCESS;
}

/*****************************************************************************
 * ItemAdded:
 *****************************************************************************/
static void ItemAdded(const vlc_event_t * event, void *user_data)
{
    services_discovery_t *sd = user_data;
    input_item_t *child = event->u.input_item_subitem_added.p_new_child;
    input_item_AddOption(child, "deinterlace=1", VLC_INPUT_OPTION_TRUSTED);
    services_discovery_AddItem(sd, child, NULL);
}


/*****************************************************************************
 * Run:
 *****************************************************************************/
static void *Run(void *data)
{
    services_discovery_t *sd = data;
    int canc = vlc_savecancel();

    const char * const name = "Freebox TV";
    const char * const url = "http://mafreebox.freebox.fr/freeboxtv/playlist.m3u";
    input_item_t *input = input_item_New(sd, url, vlc_gettext(name));
    input_item_AddOption(input, "no-playlist-autostart", VLC_INPUT_OPTION_TRUSTED);

    /* Read every subitems, and add them in ItemAdded */
    vlc_event_manager_t *em = &input->event_manager;
    vlc_event_attach(em, vlc_InputItemSubItemAdded, ItemAdded, sd);
    input_Read(sd, input);
    vlc_event_detach(em, vlc_InputItemSubItemAdded, ItemAdded, sd);

    vlc_gc_decref(input);

    vlc_restorecancel(canc);
    return NULL;
}

/*****************************************************************************
 * Close:
 *****************************************************************************/
static void Close(vlc_object_t *this)
{
    services_discovery_t *sd = (services_discovery_t *)this;
    services_discovery_sys_t *sys = sd->p_sys;
    vlc_join(sys->thread, NULL);
    free(sys);
}

static int vlc_sd_probe_Open(vlc_object_t *obj)
{
    VLC_UNUSED(obj);
    return VLC_PROBE_CONTINUE;
}

File: share/lua/sd/freebox.lua
--[[
 $Id$

 Copyright © 2010 VideoLAN and AUTHORS

 Authors: Fabio Ritrovato <sephiroth87 at videolan dot org>

 This program is free software; you can redistribute it and/or modify
 it under the terms of the GNU General Public License as published by
 the Free Software Foundation; either version 2 of the License, or
 (at your option) any later version.

 This program is distributed in the hope that it will be useful,
 but WITHOUT ANY WARRANTY; without even the implied warranty of
 MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 GNU General Public License for more details.

 You should have received a copy of the GNU General Public License
 along with this program; if not, write to the Free Software
 Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston MA 02110-1301, USA.
--]]

function descriptor()
    return { title="Freebox TV" }
end

function main()
    local logos = {}

    -- fetch urls for basic channels logos
    local fd, msg = vlc.stream( "http://free.fr/adsl/pages/television/services-de-television/acces-a-plus-250-chaines/bouquet-basique.html" )
    if not fd then
        vlc.msg.warn(msg)
        -- not fatal
    else
        local channel, logourl
        local line = fd:readline()
        while line ~= nil do
            if( string.find( line, "tv%-chaine%-" ) ) then
                _, _, channel, logourl = string.find( line, "\"tv%-chaine%-(%d+)\".*<img%s*src=\"([^\"]*)\"" )
                -- fix spaces
                logourl = string.gsub( logourl, " ", "%%20" )
                logos[channel] = "http://free.fr" .. logourl
            end
            line = fd:readline()
        end
    end

    -- fetch urls for optional channels logos
    local fd, msg = vlc.stream( "http://www.free.fr/adsl/pages/television/services-de-television/acces-a-plus-250-chaines/en-option.html" )
    if not fd then
        vlc.msg.warn(msg)
        -- not fatal
    else
        local channel, logourl
        local line = fd:readline()
        while line ~= nil do
            if( string.find( line, "tv%-chaine%-" ) ) then
                _, _, channel, logourl = string.find( line, "\"tv%-chaine%-(%d+)\".*<img%s*src=\"([^\"]*)\"" )
                -- fix spaces
                logourl = string.gsub( logourl, " ", "%%20" )
                logos[channel] = "http://free.fr" .. logourl
            end
            line = fd:readline()
        end
    end

    -- fetch the playlist
    fd, msg = vlc.stream( "http://mafreebox.freebox.fr/freeboxtv/playlist.m3u" )
    if not fd then
        vlc.msg.warn(msg)
        return nil
    end
    local line=  fd:readline()
    if line ~= "#EXTM3U" then
        return nil
    end
    line = fd:readline()
    local duration, artist, name, arturl
    local options={"deinterlace=1"}
    while line ~= nil do
        if( string.find( line, "#EXTINF" ) ) then
            _, _, duration, artist, name = string.find( line, ":(%w+),(%w+)%s*-%s*(.+)" )
            arturl = logos[artist]
        elseif( string.find( line, "#EXTVLCOPT" ) ) then
            _, _, option = string.find( line, ":(.+)" )
            table.insert( options, option )
        else
            vlc.sd.add_item( {path=line,duration=duration,artist=artist,title=name,arturl=arturl,options=options} )
            duration = nil
            artist = nil
            name = nil
            arturl = nil
            options={"deinterlace=1"}
        end
        line = fd:readline()
    end
end
